### PR TITLE
Unifying assesment

### PR DESCRIPTION
**Pull request by Gordonei**
# Overview
Pulls tasks into a single coherent whole, with different sections to be completed, based upon the different positions.

The idea is that even if being asked to work on this for a weekend, a candidate applying for two of the positions should have a reasonable chance of trying everything.

# Testing
## Confirming web assets in place
`city-hex-polygons-8-10.geojson`:
```
$ curl -I https://cct-ds-code-challenge-input-data.s3.af-south-1.amazonaws.com/city-hex-polygons-8-10.geojson
HTTP/1.1 200 OK
x-amz-id-2: i9PY/46/T51uNPItty0YjLKcbFhXKosx/izXS7z3DZRCTV1KDevK9iKq0zrJAfmJ26wsEoxD0dw=
x-amz-request-id: 2CZCFYHA9E014FY3
Date: Fri, 14 May 2021 10:47:08 GMT
Last-Modified: Fri, 14 May 2021 09:44:09 GMT
ETag: "1b257eed30f40a4645b4021567fcf0b7-7"
x-amz-storage-class: INTELLIGENT_TIERING
Accept-Ranges: bytes
Content-Type: application/geo+json
Content-Length: 108254980
Server: AmazonS3
```

`cct-ds-code-challenge-output-data` (writeonly bucket):
```
$ curl -X PUT -T s3_test_output.json https://cct-ds-code-challenge-output-data.s3.af-south-1.amazonaws.com/
$ curl https://cct-ds-code-challenge-output-data.s3.af-south-1.amazonaws.com/
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>AccessDenied</Code><Message>Access Denied</Message><RequestId>DFFN7XMJDJMVX07F</RequestId><HostId>zSR9BD7R8AfKGWE5wRXxzOS+MeQttgv5yPY5zCDSTaftCBS28+RhPma28KVb6Qm0n/Pjb4rfCN8=</HostId></Error>
$ aws s3 ls cct-ds-code-challenge-output-data
2021-05-14 12:48:06        435 s3_test_output.json
```

## Confirming S3 select works as expected
Test script:
```bash
#!/usr/bin/env bash
set -e

aws s3api select-object-content \
    --bucket "cct-ds-code-challenge-input-data" \
    --key "city-hex-polygons-7.geojson" \
    --expression "select * from s3object[*].features[*] s where s.properties.index = '87ad36180ffffff'" \
    --expression-type 'SQL' \
    --input-serialization '{"JSON": { "Type": "DOCUMENT" }, "CompressionType": "NONE"}' \
    --output-serialization '{"JSON": {}}' "s3_test_output.json"

cat s3_test_output.json
```

Testing:
```
$ bash s3_select_example.bash 
{"type":"Feature","properties":{"index":"87ad36180ffffff"},"geometry":{"type":"Polygon","coordinates":[[[1869026825870867e-14,-33867033197513507e-15],[18691689809315491e-15,-33855117645115939e-15],[18679265404549795e-15,-33847511224748857e-15],[18665420044866909e-15,-33851820194073497e-15],[18663997091762614e-15,-33863735108965621e-15],[18676420900551488e-15,-33871341692279181e-15],[1869026825870867e-14,-33867033197513507e-15]]]}}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,7 +19,7 @@ So, follow common conventions with respect to directory structure and names to m
 ### Candidates where programming is required (Data Scientist and Engineers)
 Requirements and notes:
 * Our primary programming languages are `python` and `R`. We will accept code that is packaged in `py`, `.ipynb`, `.R` and `.Rmd` files. 
-* Bash scripts are fine for glue. You may develop in any development environment you choose. 
+* Bash or similar scripting language files are fine for glue. You may develop in any development environment you choose. 
 * We expect to be able to clone your repo, immediately identify what script to execute from your README file, and execute it to completion with no human interaction. 
   In order to ensure that our environment has the right libraries or packages, please follow standard python (PEP8) or R guidelines for structure in your code, i.e place `import` and `library()` commands at the top of your scripts.
```
Comment (riazarbi) [inline README.md:24]: line 29 - change from *NB* to **NOTE** just for formatting consistency reasons

File: README.md [hunk 2]
```diff
@@ -45,55 +48,69 @@ NOTE: If you would like to _improve_ the content of this repository, by fixing t
...
+### 1. Data Extraction (if applying for Data Engineering Positions)
+We have made two resources available remotely:
+* A GeoJSON file that contains the level 8, 9 and 10 resolution hexagons for the City of Cape Town at [this location](insert readonly object url).
+* An AWS S3 writeonly bucket at [this location](Insert here).
 
-1. *Time series challenge*: Predict the weekly number of expected service requests per small area for the next 4 weeks.
-2. *Introspection challenge*: Reshape the data into number of requests per small area in the last 12 months. Augment this data with any data you may find that is relevant. Predict the number of requests from this augmented data. Determine the drivers of requests of that particular type.
-3. *Classification challenge*: Classify a small area layer as formal, informal or rural based on the data derived from the service request data. NOTE TO CCT: We'll need to pre-classify the SALs.
+Using [AWS S3 SELECT](https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-glacier-select-sql-reference-select.html) syntax to read in H3 resolution 8 data from this file. Use the `city-hex-polygons-8.geojson` file to validate your work.
```
Comment (riazarbi) [inline README.md]: Reword: Using to use

File: README.md [hunk 3]
```diff
@@ -45,55 +48,69 @@ NOTE: If you would like to _improve_ the content of this repository, by fixing t
...
+Join the file `city-hex-polygons-8.geojson` to the service request dataset, such that each service request is assigned to a single H3 hexagon. Use the `sr_hex.csv` file to validate your work.
 
-### If you are interviewing for a Data Analyst role
+For any requests where the `Latitude` and `Longitude` fields are empty, set the index value to `0`.
 
-In the `data` directory you will find a file called `data/sr_sal.csv`. It contains 36 months of service request data, where each row is a service request. A service request is a request from one of our residents to undertake significant work. The column titled 'SAL' contains the name of the Census 2011 Small Area that the particular request falls under.
+Include logging that lets the executor know how many of the records failed to join, and include a join error threshold above which the script will error out. Please also log the time taken to perform the operations described, and within reason, try to optimise latency and computational resources used.
 
+### 3. Descriptive Analytic Tasks (if applying for Data Science and/or Analyst Positions)
```
Comment (riazarbi) [inline README.md]: I think this should just be data analyst role; it'll take too much time for data science.

Comment (riazarbi) [reply README.md]: Also, does the dataset fit into an Excel spreadsheet? 1,048,576 rows

Then again, maybe that's a nice thing to test. I can think of 2 ways to deal with it in Excel - external file refs and power queries. But they are niche.

File: README.md [hunk 4]
```diff
@@ -45,55 +48,69 @@ NOTE: If you would like to _improve_ the content of this repository, by fixing t
...
 1. Provide a visual answer to the question "which areas and request types should Electricity concentrate on to reduce the overall volume of their requests".
-2. Provide a working prototype dashboard for monitoring progress in reducing Electricity service request volume per area and per type.
+2. Provide a working prototype dashboard for monitoring progress in reducing Electricity service request volume per area, and per type.
  
-An Executive-level person should be able to read this report and follow your analysis without guidance. 
+An Executive-level person should be able to read this report and follow your analysis without guidance.
 
-### If you are interviewing for a Data Engineer role
+### 4. Predictive Analytic Tasks (if applying for Data Science Positions)
+Please chose two of the following:
```
Comment (riazarbi) [inline README.md:92]: Do we want to use all request types, or limit it to a particular request group? Could limit to electricity as above.

File: README.md [hunk 5]
```diff
@@ -45,55 +48,69 @@ NOTE: If you would like to _improve_ the content of this repository, by fixing t
...
+1. *Time series challenge*: Predict the weekly number of expected service requests per hex for the next 4 weeks.
+2. *Introspection challenge*: Reshape the data into number of requests, per type, per hex in the last 12 months. Chose a particular request type, or group of requests. Develop a model that predicts the number of requests of your selected type, using the rest of your data. Based upon the model, and any other analysis, determine the drivers of requests of that particular type(s).
+3. *Classification challenge*: Classify a hex as formal, informal or rural based on the data derived from the service request data.
 
-In the `data` directory you will find a dataset file, `data/sr.csv`, which contains 36 months of service request data, where each row is a service request. A service request is a request from one of our residents to undertake significant work.
+Feel free to use any other data you can find in the public domain, except for task (3).
```
Comment (riazarbi) [inline README.md:97]: Not sure if there is enough information in the dataset for classification. Then again, these applicants are probably smarter than me.

File: README.md [hunk 6]
```diff
@@ -45,55 +48,69 @@ NOTE: If you would like to _improve_ the content of this repository, by fixing t
...
 
-We have made two resources available remotely:
-* A GeoJSON file that contains the level 8, 9 and 10 resolution hexagons for the City of Cape Town at [this location](insert readonly object url).
-* An AWS S3 writeonly bucket at [this location](Insert here).
+**The final output of the execution of your code should be a self-contained `html` file or executed `ipynb` file that is your report.** 
+ 
+A statistically minded layperson should be able to read this report and follow your analysis without guidance.https://h3geo.org/
```
Comment (riazarbi) [inline README.md]: Orphan link? Remove?

File: README.md [hunk 7]
```diff
@@ -45,55 +48,69 @@ NOTE: If you would like to _improve_ the content of this repository, by fixing t
...
+### 5. Further Data Transformations (if applying Data Engineering Positions)
+Write a script which anonymises the `sr.csv` file, but preserves the following resolutions (You may use H3 indexes or lat/lon coordinates for your spatial data):
    * location accuracy to within approximately 500m 
    * temporal accuracy to within 6 hours
    * scrubs any columns which may contain personally identifiable information.
- 
-Your code should be well formatted according to generally accepted style guides and include whatever is necessary for a team-mate unfamiliar with it to maintain it.
 
-## Contact
+We expect in the accompanying report that follows you will justify as to why this data is now anonymised.
```
Comment (riazarbi) [inline README.md]: Please limit this commentary to less than 500 words. If your code is written in a code notebook such as Jupyter notebook or Rmarkdown, you can include the report in your notebook.

**Review by riazarbi — COMMENTED** [review]
Alright, done. Added nits.

**Comment by demainou** [comment]
The final output of your analysis should be either a self-contained html file, executed ipynb file, Excel document with appropriate formatting and pivot tables, or a PowerBI file.
Change To:
You can use any tool to produce the output, e.g. Excel, Power BI, Tableau, etc.  The final deliverable needs to be a pdf report with your analysis.

File: README.md (hunk 3 above)
Comment (Gordonei) [reply README.md]: Ergh, good point. How about we ask them to only look at the last 2 months of data. Nice data munging curve ball.

File: README.md (hunk 5 above)
Comment (Gordonei) [reply README.md:97]: I can think of a few ways they can do it, based on the request type descriptions and volumes.

**Comment by Gordonei** [comment]
> The final output of your analysis should be either a self-contained html file, executed ipynb file, Excel document with appropriate formatting and pivot tables, or a PowerBI file.
> Change To:
> You can use any tool to produce the output, e.g. Excel, Power BI, Tableau, etc. The final deliverable needs to be a pdf report with your analysis.

Done!

**Comment by demainou** [comment]
In the data directory you will the following datasets:
Change To:
Use this ................. (link) to download the data
The dataset is to big to load into GitHub (~150Mb)

File: README.md (hunk 5 above)
Comment (Gordonei) [reply README.md:97]: My concern with using external data is that someone could just find an urbanisation index, and just apply that. While demonstrates creativity, it won't give us much inside into their modeling ability

**Comment by Gordonei** [comment]
> In the data directory you will the following datasets:
> Change To:
> Use this ................. (link) to download the data
> The dataset is to big to load into GitHub (~150Mb)

@demainou  - please comment on the code itself. Then I can resolve the comments as I address them.

**Comment by demainou** [comment]
sr.csv contains 36 months of service request data, where each row is a service request. A service request is a request from one of the residents of the City of Cape Town to undertake significant work. This is an important source of information on service delivery, and our performance thereof.
Change To:
sr.csv contains 36 months of service request data, where each row is a service request. A service request is a request from one of the residents of the City of Cape Town to undertake significant work. This is an important source of information on service delivery, and our performance thereof.
There is also a 12-month dataset if not able to work with the 36 month dataset

Add an additional dataset (one year) for data analyst using excel (1mil limit)

File: README.md (hunk 4 above)
Comment (Gordonei) [reply README.md:92]: Let's do Water and Sanitation to ensure a good `N`.

Comment (Gordonei) [reply README.md:89]: Let's do Water and Sanitation to ensure a good `N`.

File: README.md [hunk 8]
```diff
@@ -45,55 +48,69 @@ NOTE: If you would like to _improve_ the content of this repository, by fixing t
...
+
+For all roles, we expect the challenge response to include what you consider to be role-appropriate testing and validation. For example, a Data Scientist might want to include MAPE scores or confusion matrices. A Data Engineer may want to include logging and data quality validation tests. A Data Analyst might want to plot histograms of request counts to ensure that outliers aren't overwhelming your analysis.
+
+Your code should be well formatted according to generally accepted style guides and include whatever is necessary for a team-mate unfamiliar with it to maintain it.
 
-For all roles, we expect the challenge response to include what you consider to be role-appropriate testing and validation. For instance, a Data Scientist might want to include MAPE scores or confusion matrices. A Data Engineer may want to include logging and data quality validation tests. A Data Analyst might want to plot histograms of request counts to ensure that outliers aren't overwhelming your analysis.
+### 0. Setup
+In the `data` directory you will the following datasets:
```
Comment (demainou) [inline README.md]: Use this ................. (link) to download the data
The dataset is to big to load into GitHub (~150Mb)

File: README.md [hunk 9]
```diff
@@ -45,55 +48,69 @@ NOTE: If you would like to _improve_ the content of this repository, by fixing t
...
+* `sr.csv` contains 36 months of service request data, where each row is a service request. A service request is a request from one of the residents of the City of Cape Town to undertake significant work. This is an important source of information on service delivery, and our performance thereof.
```
Comment (demainou) [inline README.md]: sr.csv contains 36 months of service request data, where each row is a service request. A service request is a request from one of the residents of the City of Cape Town to undertake significant work. This is an important source of information on service delivery, and our performance thereof.
There is also a 12-month dataset if not able to work with the 36 month dataset

Add an additional dataset (one year) for data analyst using excel (1mil limit)

**Comment by Gordonei** [comment]
> sr.csv contains 36 months of service request data, where each row is a service request. A service request is a request from one of the residents of the City of Cape Town to undertake significant work. This is an important source of information on service delivery, and our performance thereof.
> Change To:
> sr.csv contains 36 months of service request data, where each row is a service request. A service request is a request from one of the residents of the City of Cape Town to undertake significant work. This is an important source of information on service delivery, and our performance thereof.
> There is also a 12-month dataset if not able to work with the 36 month dataset
> 
> Add an additional dataset (one year) for data analyst using excel (1mil limit)

Done

File: README.md [hunk 10]
```diff
@@ -16,17 +19,16 @@ So, follow common conventions with respect to directory structure and names to m
...
 * If your repo does not clone and run, we will not attempt to fix it.
 * If your analysis makes use of any external data, the data must either be included in the repo, or be downloaded automatically during script execution.
 
 ### Candidates where programming is not required (Data Analysts)
-*NB* If you prefer, you may submit using the requirements described above.
+*Note* If you prefer, you may submit using the requirements described above.
 
-The final output of your analysis should be either a self-contained `html` file, executed `ipynb` file, 
-Excel document with appropriate formatting and pivot tables, or a PowerBI file.
+You can use any tool to produce the output, e.g. Excel, Power BI, Tableau, etc. The final deliverable needs to be a pdf report with your analysis.
```
Comment (demainou) [inline README.md]: can we include python and R as examples?  You can use any tool to produce the output, e.g. Excel, Power BI, Tableau, Python, R etc. ......................

**Pull request merged**